lib/common/output_html.c

```diff
@@ -0,0 +1,296 @@
+
+static const char *stylesheet_default =
+    ".bold { font-weight: bold }\n"
+    ".warning { color: red, font-weight: bold }";
+
+typedef struct html_private_s {
+    xmlNode *root;
+    GQueue *parent_q;
+} html_private_t;
+
+static pcmk__output_option_t html_options[] = {
+    { "cgi-output",      "0",  "Add text needed to use output in a CGI program" },
+    { "meta-refresh",    "5",  "How often to refresh (in seconds)" },
+    { "stylesheet-link", "",   "Add a link to an external CSS stylesheet" },
+
+    { NULL, NULL, NULL }
+};
+
+static void
+html_free_priv(pcmk__output_t *out) {
+    html_private_t *priv = out->priv;
+
+    if (priv == NULL) {
+        return;
+    }
+
+    xmlFreeNode(priv->root);
+    g_queue_free(priv->parent_q);
+    free(priv);
+}
+
+static bool
+html_init(pcmk__output_t *out) {
+    html_private_t *priv = NULL;
+    htmlNodePtr body_node = NULL;
+
+    /* If html_init was previously called on this output struct, just return. */
+    if (out->priv != NULL) {
+        return true;
+    } else {
+        out->priv = calloc(1, sizeof(html_private_t));
+        if (out->priv == NULL) {
+            return false;
+        }
+
+        priv = out->priv;
+    }
+
+    for (int i = 0; html_options[i].opt_name != NULL; i++) {
+        pcmk__output_set_option(out, html_options[i].opt_name, html_options[i].opt_default);
+    }
+
+    priv->parent_q = g_queue_new();
+
+    priv->root = create_xml_node(NULL, "html");
+    g_queue_push_tail(priv->parent_q, priv->root);
+
+    body_node = xmlNewChild(g_queue_peek_tail(priv->parent_q), NULL, (pcmkXmlStr) "body", NULL);
+    g_queue_push_tail(priv->parent_q, body_node);
+
+    return true;
+}
+
+static void
+html_finish(pcmk__output_t *out, crm_exit_t exit_status) {
+    html_private_t *priv = out->priv;
+    htmlNodePtr head_node = NULL;
+    char *stylesheet = NULL;
+    char *refresh = NULL;
+
+    /* If root is NULL, html_init failed and we are being called from pcmk__output_free
+     * in the pcmk__output_new path.
+     */
+    if (priv->root == NULL) {
+        return;
+    }
+
+    if (safe_str_eq(pcmk__output_get_option(out, "cgi-output"), "1")) {
+        fprintf(out->dest, "Content-Type: text/html\n\n");
+    }
+
+    /* Add the head node last - it's not needed earlier because it doesn't contain
+     * anything else that the user could add, and we want it done last to pick up
+     * any options that may have been given.
+     */
+    head_node = xmlNewNode(NULL, (pcmkXmlStr) "head");
+    if (out->request != NULL) {
+        xmlNewTextChild(head_node, NULL, (pcmkXmlStr) "title", (pcmkXmlStr) out->request);
+    }
+
+    refresh = pcmk__output_get_option(out, "meta-refresh");
+    if (refresh != NULL) {
+        htmlNodePtr refresh_node = xmlNewChild(head_node, NULL, (pcmkXmlStr) "meta", NULL);
+        xmlSetProp(refresh_node, (pcmkXmlStr) "http-equiv", (pcmkXmlStr) "refresh");
+        xmlSetProp(refresh_node, (pcmkXmlStr) "content", (pcmkXmlStr) refresh);
+    }
+
```

Comment:
I would like to be able to support arbitrary meta options here, but the nature of having a hash table means I can't easily store a 3-tuple ("meta", name,  value).  Also it would be weird for everything that's not a meta option.  One idea I had was just iterating over all the key/value pairs in the hash table and writing out any whose names are "meta-<real name>".  Seems gross.

Reply:
Btw. there's an established pattern of using `#` prefix to stick "meta"
label onto the undecorated keys.

Reply:
Do you have an example of where this is done?

Reply:
Sorry for being perhaps too literal about "meta", "out-of-band" would
be more appropriate, I've meant these for instance:

https://clusterlabs.org/pacemaker/doc/en-US/Pacemaker/2.0/html-single/Pacemaker_Explained/index.html#_special_node_attributes

or perhaps even

https://github.com/ClusterLabs/pacemaker/blob/Pacemaker-2.0.2/lib/common/acl.c#L665

That is, something that is normally forbidden by the lexical scope in
given namespace (resolving hence the problem of mixing kinds of value
into single bucket).

Reply:
These aren't pacemaker values, they're standard HTML. E.g.

    <!DOCTYPE html>
    <html lang="en">
      <head>
        <meta charset="utf-8">
    ...

Reply:
Oh, I was looking at the purely technical aspect from the original
comment, thinking that it's just some intermediate step in the whole
flow that requires combining two namespaces into one (I've also come
by `hash2metafield` aligned with such needs recently).